### PR TITLE
Implement context-based access control

### PR DESCRIPTION
**Pull request by jorissteyn**
In a nutshell:

 - middleware configuration (/management/configuration) supports a new optional configuration key for SP en IdP configuration: use_pdp (defaults to false)
 - when enabled, the PDP endpoint (configured in parameters.yml) will be called
 - depending on the policy decision, one of the following happens:
   - deny
   - permit
   - permit with obligation: required loa already higher or equal to
   - permit with obligation: required loa is increased

### CODE REVIEW

File: src/Surfnet/StepupGateway/GatewayBundle/Controller/SecondFactorController.php [hunk 1]
```diff
@@ -67,6 +67,15 @@ public function selectSecondFactorForVerificationAction()
             $logger->notice(sprintf('Determined that the required Loa is "%s"', $requiredLoa));
         }
 
+        $requiredLoa = $this->getPdpService()->enforceObligatoryLoa(
```
Comment (MKodde) [inline src/Surfnet/StepupGateway/GatewayBundle/Controller/SecondFactorController.php]: Additional notice logging might be informative.

File: src/Surfnet/StepupGateway/GatewayBundle/Saml/ResponseContext.php [hunk 2]
```diff
@@ -171,6 +171,11 @@ public function saveAssertion(SAML2_Assertion $assertion)
         $authenticatingAuthorities = $assertion->getAuthenticatingAuthority();
         if (!empty($authenticatingAuthorities)) {
             $this->stateHandler->setAuthenticatingIdp(reset($authenticatingAuthorities));
+        } else {
```
Comment (MKodde) [inline src/Surfnet/StepupGateway/GatewayBundle/Saml/ResponseContext.php]: Nice addition!

File: src/Surfnet/StepupGateway/GatewayBundle/Tests/Service/PdpServiceTest.php [hunk 3]
```diff
@@ -335,4 +338,93 @@ public function exception_is_thrown_if_obligation_requires_unknown_loa()
...
+
+        $context = Mockery::mock(ResponseContext::class);
+        $context->shouldReceive('getAuthenticatingIdp')
+            ->andReturnNull();
+        $context->shouldReceive('getServiceProvider')
+            ->andReturn($sp);
+
+        $enabled = $service->isEnabledForSpOrIdp($context);
+
+        $this->assertFalse($enabled);
```
Comment (MKodde) [inline src/Surfnet/StepupGateway/GatewayBundle/Tests/Service/PdpServiceTest.php]: Fixed this by changing the expected value.

**Comment by thijskh** [comment]
Calling this PDP functonality has now been implemented in the Stepup interface in OpenConext/OpenConext-engineblock. Since the code has landed elsewhere, the underlying Stepup-Gateway code has diverged and the PDP interface has changed, this can now be closed.

**Pull request closed without merging**